feat: Add store task benchmarking with image analysis utilities.

jioffe502 opened the pull request:
Implements comprehensive benchmarking for nv-ingest store task performance comparison and configurable image extraction to disk for visual analysis.

## Description
Adds comprehensive benchmarking tools to measure the performance impact of nv-ingest's MinIO store task on throughput and latency. Includes configurable image extraction utilities that save charts, tables, and infographics to disk for detailed analysis and debugging.

Results for ablation study in slack.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/nv-ingest/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] If adjusting docker-compose.yaml environment variables have you ensured those are mimicked in the Helm values.yaml file.

VibhuJawa commented on client/src/nv_ingest_client/util/image_disk_utils.py:
```diff
@@ -0,0 +1,260 @@
+            # Decode base64 content and save as image file
+            try:
+                # Decode base64 image data
+                image_data = base64.b64decode(image_content)
+                image = Image.open(io.BytesIO(image_data))
```
We  could use `ensure_base64_format`   `from nv_ingest_api.util.image_processing.transforms` conversation, mostly faster than PIL>

jioffe502 replied:
Nice catch @VibhuJawa ! I just changed the functionality to leverage this ensure_base64_format and it is faster. 

Also, set default to jpeg and found ~8x speedup in my latest commit per your question on slack.

VibhuJawa commented on client/src/nv_ingest_client/util/image_disk_utils.py:
```diff
@@ -24,6 +24,13 @@
+    from nv_ingest_api.util.image_processing.transforms import base64_to_numpy
+
+    OPENCV_AVAILABLE = True
+except ImportError:
```
Interesting, Given this a client side fix, there are 2 ways to fix this. 

1. One we add open cv as a dependency on client (which should be fine) 

or 

2. We  fall back to PIL always, 

We should not fork this.

jioffe502 replied:
I am torn because in nv_ingest_api.util.image_processing.transforms import base64_to_numpy OpenCV is used and in: nv_ingest_client/cli/util/processing.py PIL is used but there is no dependencies for either in the client if I am looking in the right place (client/pyproject.toml)

VibhuJawa commented on client/src/nv_ingest_client/util/image_disk_utils.py:
```diff
@@ -138,33 +149,47 @@ def save_images_to_disk(
+
+                    # Convert numpy array back to PIL Image for saving
+                    # (base64_to_numpy returns RGB format)
+                    image = Image.fromarray(img_array)
```
If all we want to do is save the image to jpeg we should not create ` image = Image.fromarray(img_array)` here because that will cause a image_content->jpeg->PIL->jpeg round trip. 

I wonder if we can directly write it :

```python3
def base64_to_disk(b64_string: str, output_path: str):
    # Strip off metadata if present (e.g., "data:image/jpeg;base64,")
    if "," in b64_string:
        b64_string = b64_string.split(",")[1]

    # Decode and write directly as JPEG
    with open(output_path, "wb") as f:
        f.write(base64.b64decode(b64_string))

b64_string = ensure_base64_format(image_content, format)
base64_to_disk(b64_string)
```

Like, PIL decoding/encoding round trip is expensive and we should avoid it.

jioffe502 replied:
Gave this a shot and handled it by reading the format and performing a direct write if it matches, preventing any conversion loop.

Also, tested full opencv implementation vs full PIL implementation on Bo20. There is no difference in image processing perf, but openCV makes the full pipeline faster by ~10% 

Still up for debate based on versioning of course but I have two versions ready

randerzander replied:
I think the store_task is already likely to be slow due to storage latency (cloud, or minio)

Given the intent here is to make it easy for users to easily interact w/ extracted images, why not just make the format it uses to save to disk configurable?

Use the fast base64 route by default, and jpeg if user is fine w/ paying the cost because that's easier/preferred for them

VibhuJawa reviewed (changes requested):
We should either stick to PIL or add opencv as a dependency to client. My suggestion is adding that dependency because it is much faster in my earlier experiments

jioffe502 commented:
@VibhuJawa @drobison00 reduced duplicate code by moving necessary utilities into transforms.py on the api side and made a few new additions to transforms that include unit tests. Additionally, I simplified some of the logic on the client side when it came to image format processing. All using opencv now from the api side so no new dependencies needed.

Also, this PR establishes the foundation for future enhancements to the .store() task. The new base64_to_disk() and save_image_to_disk() functions in transforms.py provide reusable, well-tested components that can be leveraged when we add fsspec support to enable cloud storage (S3, GCS, Azure) in the ingestion pipeline. I also have benchmark baselines of how our current store task is impacting performance. 

Next steps is a future PR that will integrate fsspec into the .store() task for multi-cloud storage or disk storage during ingestion

edknv commented on scripts/tests/cases/store_benchmark.py:
```diff
@@ -0,0 +1,716 @@
+            Ingestor: Configured ingestor instance ready for ingestion
+        """
+        # Set up temporary file directory in /raid instead of /tmp
+        temp_dir = f"/raid/jioffe/tmp/store_task/{self.run_id}_{collection_name}"
```
This should be generalized and also configurable.

edknv commented on scripts/tests/cases/store_benchmark.py:
```diff
@@ -0,0 +1,716 @@
+        try:
+            import shutil
+
+            temp_dir = f"/raid/jioffe/tmp/store_task/{self.run_id}_{collection_name}"
+            image_dir = f"/raid/jioffe/tmp/store_task/{self.run_id}_{collection_name}_images"
```
Related to the previous comment, if `temp_dir` is changed because user is able to set it different, it should change here too.

edknv commented on scripts/tests/cases/store_benchmark.py:
```diff
@@ -0,0 +1,716 @@
+                if store_enabled and not skip_image_analysis:
+                    try:
+                        # Save images to disk for detailed analysis
+                        image_output_dir = f"/raid/jioffe/tmp/store_task/{self.run_id}_{collection_name}_images"
```
Make this configurable?

edknv commented on scripts/tests/cases/store_benchmark.py:
```diff
@@ -0,0 +1,716 @@
+        print(f"Running {condition_name} benchmark for {self.dataset_name}")
+        print(f"Store enabled: {store_enabled}")
+        print(f"Collection: {collection_name}")
+        temp_path = f"/raid/jioffe/tmp/store_task/{self.run_id}_{collection_name[:20]}..."
```
Does this need to hard coded here?

VibhuJawa reviewed (changes requested):
Most changes to the core codebase look good, added nits. 

We  should clean up the `scripts/tests/cases` a bit to ensure its generalized and cleaned up.

jioffe502 commented on client/src/nv_ingest_client/util/image_disk_utils.py:
```diff
@@ -0,0 +1,269 @@
+            target_format = output_format.lower()
+
+            # Determine file extension for naming - API handles format conversion
+            if target_format == "auto":
+                file_ext = "jpg"  # Default extension when preserving original format
```
did this with new: test_auto_format_preserves_original_png_vs_jpeg

jioffe502 commented:
Thanks for the reviews @VibhuJawa @edknv 

I went ahead and took out scripts_benchmarks and put into my private local. I came to the conclusion that It shouldn't be included as a core test in nv-ingest and something that I can benchmark privately. 

I also made some logic changes with the format preservation, simplified it, and added unit tests. Also tested this with my store_benchmark to make sure it works in real scenarios.

VibhuJawa reviewed (approved):
Minor nit around logging, looks good other wise
